livecheck and bump: formula/cask disambiguation

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----

When running `livecheck` without the `--json` option, or `bump`, formulae and casks with the same name are indistinguishable in the output even when using `--full-name`, if they are from `homebrew-core` and `homebrew-cask`.

Example:

```
➜ brew livecheck volt homebrew/cask/volt --full-name
Warning: Treating volt as a formula. For the cask, use homebrew/cask/volt
volt : 0.3.7 ==> 0.3.7
volt : 0.80 ==> 0.87
```

It would be desirable to disambiguate names in such cases, unless the `--cask` option is explicitly given.

```
➜ brew livecheck homebrew/cask/volt volt                    
Warning: Treating volt as a formula. For the cask, use homebrew/cask/volt
volt (cask) : 0.80 ==> 0.87
volt : 0.3.7 ==> 0.3.7
```

I'm not sure if handling this in `livecheck` or `bump` is ideal, maybe we would want a method in `cask` instead?

Bo98 commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -161,11 +161,18 @@ def retrieve_and_display_info(formula_or_cask, name, repositories, args:)
 
+    print_name = begin
+      "#{name} (cask)" if formula_or_cask.is_a?(Cask::Cask) && Formula[name] && !args.cask?
+    rescue
```
I know this isn't the only instance of this, but I'd try make this more specific, like `rescue FormulaUnavailableError`.

nandahkrishna replied:
Changed this and the other instance to be specific.

samford reviewed (dismissed):
This is a pain point I've encountered before and I agree that addressing it makes the default output more understandable when there's a name conflict.

One issue I'm seeing is that `(cask)` is still printed after the name if you check a cask where a formula exists with the same name but don't include the formula in the run. For example, `brew livecheck homebrew/cask/volt` prints `volt (cask) : 0.80 ==> 0.87`. I imagine we would only want to display `(cask)` when it's necessary to distinguish between multiple entries with the same name in a given run.

Addressing this would require us to identify if a given livecheck run includes both a formula and cask with the same name, instead of simply checking for the existence of a formula with the same name regardless of whether it's part of the run.

With this in mind, we may also want to identify if more than one formula in a given run has the same name. For example, if we have 2+ formulae in different taps with the same name, we can't tell these apart in the default output unless `--full-name` is used. We could potentially handle those by automatically printing the full formula name (e.g., `a/b/example`) for non-homebrew/core taps, even when `--full-name` isn't used.

What do you think about this?

nandahkrishna commented:
> With this in mind, we may also want to identify if more than one formula in a given run has the same name. For example, if we have 2+ formulae in different taps with the same name, we can't tell these apart in the default output unless --full-name is used.

As you'd suggested, I think we should be printing the full name (for just those formulae/casks with the same name) in this case as well.

> I imagine we would only want to display (cask) when it's necessary to distinguish between multiple entries with the same name in a given run.

That's right, I'll add a check for this and also the case where we have 2+ formulae/casks of the same name from different taps.

nandahkrishna commented:
I've improved the way this is handled based on the suggestions.

One question I had was whether we should disambiguate formulae for `brew bump`, because that doesn't really work well with external taps (if I'm not wrong, Repology provides data only for homebrew-core and homebrew-cask). I've added the `--full-name` flag and the code to handle this, but now I think it probably should be removed.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
-    else
-      args.named.to_formulae_and_casks.presence
-    end
+    formulae_and_casks =
+      if args.formula?
```
```suggestion
    formulae_and_casks = if args.formula?
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
+        args.named.to_casks
+      else
+        args.named.to_formulae_and_casks
+      end&.sort_by do |formula_or_cask|
```
`end&` is a bit weird; I'd split this out into a variable.

nandahkrishna replied:
I'd actually based this part of the code on [`dev-cmd/livecheck.rb`](https://github.com/Homebrew/brew/blob/5fed5536f4cf35aa86e300cb72fc5c4194b17587/Library/Homebrew/dev-cmd/livecheck.rb#L99), so should that be changed too?

MikeMcQuaid replied:
Ideally, yeh!

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
+      if !args.formula? && !args.cask?
+        ambiguous_casks = formulae_and_casks
+                          .group_by { |item| Livecheck.formula_or_cask_name(item, full_name: true) }
+                          .select { |_name, items| items.length > 1 }
```
```suggestion
                          .select { |_, items| items.length > 1 }
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
 
+      ambiguous_casks = []
+      if !args.formula? && !args.cask?
+        ambiguous_casks = formulae_and_casks
+                          .group_by { |item| Livecheck.formula_or_cask_name(item, full_name: true) }
```
```suggestion
        ambiguous_casks = formulae_and_casks.group_by { |item| Livecheck.formula_or_cask_name(item, full_name: true) }
```

feels less weird than the variable name "floating"

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
+                          .group_by { |item| Livecheck.formula_or_cask_name(item, full_name: true) }
+                          .select { |_name, items| items.length > 1 }
+                          .values.flatten
+                          .select { |item| item.is_a?(Cask::Cask) }
```
If you rebase this on `master` I think `rubocop` will complain about this and auto fix to a better version.

nandahkrishna replied:
Did this, don't think there were any style errors.

MikeMcQuaid replied:
np, my mistake.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
+
+      ambiguous_names = []
+      unless args.full_name?
+        ambiguous_names = (formulae_and_casks - ambiguous_casks)
+                          .group_by { |item| Livecheck.formula_or_cask_name(item) }
```
```suggestion
        ambiguous_names = (formulae_and_casks - ambiguous_casks).group_by { |item| Livecheck.formula_or_cask_name(item) }
```

feels less weird than the variable name "floating"

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -105,9 +134,14 @@ def bump
           end
           name = Livecheck.formula_or_cask_name(formula_or_cask)
+          ambiguous_cask = begin
+            formula_or_cask.is_a?(Cask::Cask) && !args.cask? && Formula[name] ? true : false
```
```suggestion
           (formula_or_cask.is_a?(Cask::Cask) && !args.cask? && Formula[name]).present?
```

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -97,10 +98,28 @@ def load_other_tap_strategies(formulae_and_casks_to_check)
 
+      ambiguous_casks = []
+      if handle_name_conflict
+        ambiguous_casks = formulae_and_casks_to_check
+                          .group_by { |item| formula_or_cask_name(item, full_name: true) }
```
```suggestion
        ambiguous_casks = formulae_and_casks_to_check.group_by { |item| formula_or_cask_name(item, full_name: true) }
```

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -97,10 +98,28 @@ def load_other_tap_strategies(formulae_and_casks_to_check)
+
+      ambiguous_names = []
+      unless full_name
+        ambiguous_names = (formulae_and_casks_to_check - ambiguous_casks)
+                          .group_by { |item| formula_or_cask_name(item) }
```
```suggestion
        ambiguous_names = (formulae_and_casks_to_check - ambiguous_casks).group_by { |item| formula_or_cask_name(item) }
```

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -97,10 +98,28 @@ def load_other_tap_strategies(formulae_and_casks_to_check)
+        ambiguous_casks = formulae_and_casks_to_check
+                          .group_by { |item| formula_or_cask_name(item, full_name: true) }
+                          .select { |_name, items| items.length > 1 }
+                          .values.flatten
```
```suggestion
                          .values
                          .flatten
```

but, given you don't use the values for the `.select` you could use `.values` before that.

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -97,10 +98,28 @@ def load_other_tap_strategies(formulae_and_casks_to_check)
+        ambiguous_names = (formulae_and_casks_to_check - ambiguous_casks)
+                          .group_by { |item| formula_or_cask_name(item) }
+                          .select { |_name, items| items.length > 1 }
+                          .values.flatten
```
```suggestion
                          .values
                          .flatten
```

but, given you don't use the values for the `.select` you could use `.values` before that.

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -97,10 +98,28 @@ def load_other_tap_strategies(formulae_and_casks_to_check)
+      if handle_name_conflict
+        ambiguous_casks = formulae_and_casks_to_check
+                          .group_by { |item| formula_or_cask_name(item, full_name: true) }
+                          .select { |_name, items| items.length > 1 }
```
```suggestion
                          .select { |_, items| items.length > 1 }
```

MikeMcQuaid commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -97,10 +98,28 @@ def load_other_tap_strategies(formulae_and_casks_to_check)
+      unless full_name
+        ambiguous_names = (formulae_and_casks_to_check - ambiguous_casks)
+                          .group_by { |item| formula_or_cask_name(item) }
+                          .select { |_name, items| items.length > 1 }
```
```suggestion
                          .select { |_, items| items.length > 1 }
```

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
+        ambiguous_casks = formulae_and_casks
+                          .group_by { |item| Livecheck.formula_or_cask_name(item, full_name: true) }
+                          .select { |_name, items| items.length > 1 }
+                          .values.flatten
```
```suggestion
                          .values
                          .flatten
```

but, given you don't use the values for the `.select` you could use `.values` before that.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bump.rb:
```diff
@@ -38,23 +40,44 @@ def bump
+        ambiguous_names = (formulae_and_casks - ambiguous_casks)
+                          .group_by { |item| Livecheck.formula_or_cask_name(item) }
+                          .select { |_name, items| items.length > 1 }
+                          .values.flatten
```
```suggestion
                          .values
                          .flatten
```

but, given you don't use the values for the `.select` you could use `.values` before that.

MikeMcQuaid reviewed (approved):
Happy with this when you are!

nandahkrishna commented:
Thanks Mike! Just waiting for Sam's opinion on this before merging.

samford commented:
I'm aiming to go through this again sometime today or tomorrow. The main area I'm interested in is the logic that identifies ambiguous names. I tinkered with an alternative implementation yesterday which may be less intensive but I still need to do some more profiling.

MikeMcQuaid commented:
> I tinkered with an alternative implementation yesterday which may be less intensive but I still need to do some more profiling.

Please use `brew prof` or `hyperfine` for any profiling rather than anything per-hand. We're interested in the total runtime of the process rather than the runtime of specific methods within it.

nandahkrishna commented:
Some benchmarks (without disambiguation vs with), although they may not be too comprehensive:

* A simple run:
```
➜ git branch --show-current
master
➜ hyperfine -i "brew livecheck airtool python cmake homebrew/cask/cmake volt homebrew/cask/volt terraform hashicorp/tap/terraform"
Benchmark #1: brew livecheck airtool python cmake homebrew/cask/cmake volt homebrew/cask/volt terraform hashicorp/tap/terraform
  Time (mean ± σ):      9.011 s ±  0.169 s    [User: 921.9 ms, System: 702.2 ms]
  Range (min … max):    8.641 s …  9.221 s    10 runs

  Warning: Ignoring non-zero exit code.

➜ cb formula-cask-disambiguation
Switched to branch 'formula-cask-disambiguation'
➜ hyperfine -i "brew livecheck airtool python cmake homebrew/cask/cmake volt homebrew/cask/volt terraform hashicorp/tap/terraform"
Benchmark #1: brew livecheck airtool python cmake homebrew/cask/cmake volt homebrew/cask/volt terraform hashicorp/tap/terraform
  Time (mean ± σ):      8.852 s ±  0.133 s    [User: 933.2 ms, System: 716.2 ms]
  Range (min … max):    8.611 s …  9.083 s    10 runs

  Warning: Ignoring non-zero exit code.

```

* All formulae and casks with the same name (in `homebrew-core` and `homebrew-cask`):
```
➜ git branch --show-current
master
➜ hyperfine -i --max-runs=5 "brew livecheck $(cat ~/Downloads/formulae-casks-conflict.txt)"

Benchmark #1: brew livecheck angband appium archiver atlantis aurora bee blitz brightness ckan cmake cocoapods crunch cutter dash davmail docker dosbox-x dosbox dotnet doxygen emacs fetch flow flume flux fontforge foreman fs-uae gitup handbrake hey igv ipfs jasper jump kawa kite knock lab lepton lynx macvim mame marked megacmd metabase mgba mkvtoolnix mpv namebench nano neo4j nomad noti nzbget openmsx openrct2 opensc openttd osquery platypus puzzles r rabbitmq redis rhino screen scummvm silicon sip snappy spark squirrel stella subnetcalc svgcleaner swift swimat syncthing tag termius unison vapor volt wakeonlan widelands wireshark homebrew/cask/angband homebrew/cask/appium homebrew/cask/archiver homebrew/cask/atlantis homebrew/cask/aurora homebrew/cask/bee homebrew/cask/blitz homebrew/cask/brightness homebrew/cask/ckan homebrew/cask/cmake homebrew/cask/cocoapods homebrew/cask/crunch homebrew/cask/cutter homebrew/cask/dash homebrew/cask/davmail homebrew/cask/docker homebrew/cask/dosbox-x homebrew/cask/dosbox homebrew/cask/dotnet homebrew/cask/doxygen homebrew/cask/emacs homebrew/cask/fetch homebrew/cask/flow homebrew/cask/flume homebrew/cask/flux homebrew/cask/fontforge homebrew/cask/foreman homebrew/cask/fs-uae homebrew/cask/gitup homebrew/cask/handbrake homebrew/cask/hey homebrew/cask/igv homebrew/cask/ipfs homebrew/cask/jasper homebrew/cask/jump homebrew/cask/kawa homebrew/cask/kite homebrew/cask/knock homebrew/cask/lab homebrew/cask/lepton homebrew/cask/lynx homebrew/cask/macvim homebrew/cask/mame homebrew/cask/marked homebrew/cask/megacmd homebrew/cask/metabase homebrew/cask/mgba homebrew/cask/mkvtoolnix homebrew/cask/mpv homebrew/cask/namebench homebrew/cask/nano homebrew/cask/neo4j homebrew/cask/nomad homebrew/cask/noti homebrew/cask/nzbget homebrew/cask/openmsx homebrew/cask/openrct2 homebrew/cask/opensc homebrew/cask/openttd homebrew/cask/osquery homebrew/cask/platypus homebrew/cask/puzzles homebrew/cask/r homebrew/cask/rabbitmq homebrew/cask/redis homebrew/cask/rhino homebrew/cask/screen homebrew/cask/scummvm homebrew/cask/silicon homebrew/cask/sip homebrew/cask/snappy homebrew/cask/spark homebrew/cask/squirrel homebrew/cask/stella homebrew/cask/subnetcalc homebrew/cask/svgcleaner homebrew/cask/swift homebrew/cask/swimat homebrew/cask/syncthing homebrew/cask/tag homebrew/cask/termius homebrew/cask/unison homebrew/cask/vapor homebrew/cask/volt homebrew/cask/wakeonlan homebrew/cask/widelands homebrew/cask/wireshark
  Time (mean ± σ):     248.249 s ± 12.948 s    [User: 5.859 s, System: 3.807 s]
  Range (min … max):   235.802 s … 270.011 s    5 runs

  Warning: Ignoring non-zero exit code.

➜ cb formula-cask-disambiguation
Switched to branch 'formula-cask-disambiguation'
➜ hyperfine -i --max-runs=5 "brew livecheck $(cat ~/Downloads/formulae-casks-conflict.txt)"
Benchmark #1: brew livecheck angband appium archiver atlantis aurora bee blitz brightness ckan cmake cocoapods crunch cutter dash davmail docker dosbox-x dosbox dotnet doxygen emacs fetch flow flume flux fontforge foreman fs-uae gitup handbrake hey igv ipfs jasper jump kawa kite knock lab lepton lynx macvim mame marked megacmd metabase mgba mkvtoolnix mpv namebench nano neo4j nomad noti nzbget openmsx openrct2 opensc openttd osquery platypus puzzles r rabbitmq redis rhino screen scummvm silicon sip snappy spark squirrel stella subnetcalc svgcleaner swift swimat syncthing tag termius unison vapor volt wakeonlan widelands wireshark homebrew/cask/angband homebrew/cask/appium homebrew/cask/archiver homebrew/cask/atlantis homebrew/cask/aurora homebrew/cask/bee homebrew/cask/blitz homebrew/cask/brightness homebrew/cask/ckan homebrew/cask/cmake homebrew/cask/cocoapods homebrew/cask/crunch homebrew/cask/cutter homebrew/cask/dash homebrew/cask/davmail homebrew/cask/docker homebrew/cask/dosbox-x homebrew/cask/dosbox homebrew/cask/dotnet homebrew/cask/doxygen homebrew/cask/emacs homebrew/cask/fetch homebrew/cask/flow homebrew/cask/flume homebrew/cask/flux homebrew/cask/fontforge homebrew/cask/foreman homebrew/cask/fs-uae homebrew/cask/gitup homebrew/cask/handbrake homebrew/cask/hey homebrew/cask/igv homebrew/cask/ipfs homebrew/cask/jasper homebrew/cask/jump homebrew/cask/kawa homebrew/cask/kite homebrew/cask/knock homebrew/cask/lab homebrew/cask/lepton homebrew/cask/lynx homebrew/cask/macvim homebrew/cask/mame homebrew/cask/marked homebrew/cask/megacmd homebrew/cask/metabase homebrew/cask/mgba homebrew/cask/mkvtoolnix homebrew/cask/mpv homebrew/cask/namebench homebrew/cask/nano homebrew/cask/neo4j homebrew/cask/nomad homebrew/cask/noti homebrew/cask/nzbget homebrew/cask/openmsx homebrew/cask/openrct2 homebrew/cask/opensc homebrew/cask/openttd homebrew/cask/osquery homebrew/cask/platypus homebrew/cask/puzzles homebrew/cask/r homebrew/cask/rabbitmq homebrew/cask/redis homebrew/cask/rhino homebrew/cask/screen homebrew/cask/scummvm homebrew/cask/silicon homebrew/cask/sip homebrew/cask/snappy homebrew/cask/spark homebrew/cask/squirrel homebrew/cask/stella homebrew/cask/subnetcalc homebrew/cask/svgcleaner homebrew/cask/swift homebrew/cask/swimat homebrew/cask/syncthing homebrew/cask/tag homebrew/cask/termius homebrew/cask/unison homebrew/cask/vapor homebrew/cask/volt homebrew/cask/wakeonlan homebrew/cask/widelands homebrew/cask/wireshark
  Time (mean ± σ):     248.312 s ±  8.476 s    [User: 5.928 s, System: 3.714 s]
  Range (min … max):   239.823 s … 259.555 s    5 runs

  Warning: Ignoring non-zero exit code.

```

I found [some material]() indicating that using `group_by` and `select` is quite fast compared to some other methods of identifying duplicate elements. I was able to reproduce similar results in some simple benchmarks. Also, using `select` to keep just casks in `ambiguous_casks` turns out to be beneficial as it reduces the time taken for `include?` checks.

Of course, there could be faster alternatives so maybe we could merge this and open follow-up PRs for improvements.

MikeMcQuaid commented:
Performance looks fine to me (not much in it), thanks @nandahkrishna. Let's merge this an iterate if need be.
